### DISCUSSION TITLE
More than one client connected to the server

### DISCUSSION THREAD

**Discussion by LeorFinacre** [jchristn/WatsonTcp#184] [Q&A] [answered]
Hello and thank you for this excellent project. So far it suited all my needs.

I am developing an application whose main features are:

- The server starts and displays a winform
- Clients can connect to the server and dispay the same winform
- Each time a client connects / disconnects, a list of "connected clients" is updated on all screens (server and clients)
- When the server opens a window, it is also displayed on the client side on all clients connected or selected from the server.

So far I manage to do what I just described but with only one client connected. As soon as I try to connect more than one client :

1. the server blocks (bad)
2. the first client retrieves the information that there are 2 clients connected (exactly what I want on each win form)
3. the last client has no information (bad)

What am I doing wrong ? Why is my code unable to _Server.send() twice in a loop ?
Here is a snippet of the part in error : 

```
        private void OnClientConnected(object sender, ConnectionEventArgs e)
        {
            dictionnaire.Clear();
            Logger(Severity.Debug, "Client " + e.IpPort + " connected");
            this.Invoke((MethodInvoker)delegate
            {
                listClientIP.Items.Add(e.IpPort);
                clientList.Add(e.IpPort);
                data = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(clientList));
                ms = new MemoryStream(data);
                dictionnaire.Add("simpleData", "clientList");
                foreach (var c in listClientIP.Items)
                {
                    _Server.Send(c.ToString(), data.Length, ms, dictionnaire);
                }
            });
        }
```

Thanks in advance for any help you can give me. You can ask me more information if needed.

Leor

**Comment by LeorFinacre** [comment, marked as answer]
Hi again,

I just found what the problem was. I come here to post the solution so that it can be useful to someone else.

Just move the assignment row in the MemoryStream in the for each loop

```
        private void OnClientConnected(object sender, ConnectionEventArgs e)
        {
            dictionnaire.Clear();
            Logger(Severity.Debug, "Client " + e.IpPort + " connected");
            this.Invoke((MethodInvoker)delegate
            {
                listClientIP.Items.Add(e.IpPort);
                clientList.Add(e.IpPort);
                data = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(clientList));
                dictionnaire.Add("simpleData", "clientList");
                foreach (var c in listClientIP.Items)
                {
                    ms = new MemoryStream(data);
                    _Server.Send(c.ToString(), data.Length, ms, dictionnaire);
                }
            });
        }
```

Therefore, no more problems. Clients can connect and display the list of clients IP connected to the server.

Leor

**Reply by jchristn** [reply]
Thank you for sharing this!
